# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/ipc/RemoteAccessibleBase.cpp
# Commit: fda6706bb580
# Full Hash: fda6706bb580ad9ffa023daf7c39f95106678546
# Author: Morgan Rae Reschenberg <mreschenberg@mozilla.com>
# Date: 2022-10-21 08:39:18
# Regressor Bug: 1787284
# File Overlap Count: 1
# Description:
#   Bug 1787284: [Part 1] Use tree heirarchy to compute MEMBER_OF in parent process for aria radio buttons r=eeejay
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D158892
# ==============================================================================

diff -r 7b04096d6b00 -r fda6706bb580 accessible/ipc/RemoteAccessibleBase.cpp
--- a/accessible/ipc/RemoteAccessibleBase.cpp	Fri Oct 21 00:53:17 2022 +0300
+++ b/accessible/ipc/RemoteAccessibleBase.cpp	Thu Oct 20 21:40:05 2022 +0000
@@ -724,6 +724,38 @@
     return Relation();
   }
 
+  if (aType == RelationType::MEMBER_OF) {
+    Relation rel = Relation();
+    if (IsARIARole(nsGkAtoms::radio)) {
+      // ARIA radio buttons should be grouped by their radio group
+      // parent, if one exists.
+      RemoteAccessible* currParent = RemoteParent();
+      while (currParent && currParent->Role() != roles::RADIO_GROUP) {
+        currParent = currParent->RemoteParent();
+      }
+
+      if (currParent && currParent->Role() == roles::RADIO_GROUP) {
+        // If we found a radiogroup parent, search for all
+        // roles::RADIOBUTTON children and add them to our relation.
+        // This search will include the radio button this method
+        // was called from, which is expected.
+        Pivot p = Pivot(currParent);
+        PivotRoleRule rule(roles::RADIOBUTTON);
+        Accessible* match = p.Next(currParent, rule);
+        while (match) {
+          MOZ_ASSERT(match->IsRemote(),
+                     "We should only be traversing the remote tree.");
+          rel.AppendTarget(match->AsRemote());
+          match = p.Next(match, rule);
+        }
+      }
+    }
+    // By webkit's standard, aria radio buttons do not get grouped
+    // if they lack a group parent, so we return an empty
+    // relation here if the above check fails.
+    return rel;
+  }
+
   Relation rel;
   if (!mCachedFields) {
     return rel;
